### ISSUE TITLE
To env or not to env

### ISSUE THREAD

**Issue by kseistrup** [kseistrup/filters#2]
**TL;DR:** Should we rather invoke the required interpreter with `env`, or should we build the scripts such that the shebang line has actual path to the interpreter?

As seen in #1, @mkg20001 has authored a [package file for NixOS](https://github.com/mkg20001/nix/tree/master/pkgs/kseistrup-filters) with the following pre-install command:

```sh
sed "s|env -S|env|g" -i bin/*.awk
```

effectively removing the `-S` switch. Understandably so, because said switch seems to be a feature of `env` from GNU Coreutils only. Specifically, it is absent from [IEEE Std 1003.1][1], [NetBSD][2] and [OpenBSD][3].

However, [`execve(2)`][4] has the form

```sh
#! interpreter [optional-arg]
```

meaning that — at least on some systems — AWK scripts with a shebang line like

```sh
#!/usr/bin/env awk -f
```

will result in `env` calling `awk -f` in `argv[0]` and fail, which is the reason the `-S` switch in GNU Coreutils exists in the first place.

Now, if some systems have to patch the scripts to get rid of the `-S` switch, shouldn't we rather have the absolute paths of the interpreters in the `Makefile` so that the scripts are called immediately by the correct interpreter instead of the `env`-workaround (which has always felt kludgy to me anyway)?

----
[1]: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/env.html
[2]: https://netbsd.gw.com/cgi-bin/man-cgi?env+1+NetBSD-current
[3]: https://man.openbsd.org/env.1
[4]: https://linux.die.net/man/2/execve

**Comment by mkg20001** [comment]
cross-ref: https://github.com/NixOS/nixpkgs/issues/77539

**Comment by kseistrup** [comment]
In 31e6a7548ee58fb9bfd6d9a710457d12dbf444ef I have ommitted `/usr/bin/env` from the AWK scripts, and instead used the shebang path `#!/usr/bin/awk -f`.

If `awk` is not `/usr/bin/awk`, it _should_ be possible to call `make patch-awk` in order to have `sed` insert the correct shebang path. This is experimental and I am not an expert in makefiles, so…

I have kept `env` in the `bash` and `sh` scripts, since we aren't exceeding two arguments here. However, there are also `patch-bash`, `patch-sh` (and even `patch-env`) targets, should somebody wish to get rid of the `env` entirely.

**Comment by kseistrup** [comment]
Issues now live at Codeberg: https://codeberg.org/kas/filters/issues
